vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution_unblinded.cc

```diff
@@ -131,176 +114,405 @@ std::string GenerateTokenPayload(
+      ledger::ContributionStep::STEP_SUGGESTIONS,
+      retry_count,
+      [](const ledger::Result){});
+
+  for (auto& publisher : contribution->publishers) {
+    if (publisher->total_amount == publisher->contributed_amount) {
       continue;
     }
 
-    ledger::ReconcileDirectionProperties direction;
-    direction.publisher_key = winner.first;
-    direction.amount_percent = (winner.second / total_votes) * 100;
-    new_directions.push_back(direction);
+    std::vector<ledger::UnblindedToken> token_list;
+    double current_amount = 0.0;
+    for (auto& item : list) {
+      if (current_amount >= publisher->total_amount) {
+        break;
+      }
 
-    const uint32_t new_position = current_position + winner.second;
-    ledger::UnblindedTokenList new_list;
-    for (size_t i = current_position; i < new_position; i++) {
-      new_list.push_back(std::move(list[i]));
+      current_amount += item.value;
+      token_list.push_back(item);
     }
-    current_position = new_position;
+
+    auto callback = std::bind(&Unblinded::TokenProcessed,
+        this,
+        _1,
+        contribution->contribution_id,
+        publisher->publisher_key);
 
     SendTokens(
-        winner.first,
-        ledger::RewardsType::AUTO_CONTRIBUTE,
-        std::move(new_list),
-        [](const ledger::Result _){});
+        publisher->publisher_key,
+        contribution->type,
+        token_list,
+        callback);
+    return;
   }
 
-  reconcile.directions = new_directions;
-  ledger_->UpdateReconcile(reconcile);
+  ContributionCompleted(ledger::Result::LEDGER_OK, std::move(contribution));
+}
+
+void Unblinded::TokenProcessed(
+    const ledger::Result result,
+    const std::string& contribution_id,
+    const std::string& publisher_key) {
+  if (result == ledger::Result::LEDGER_OK) {
+    auto callback = std::bind(&Unblinded::OnTokenProcessed,
+        this,
+        _1,
+        contribution_id);
 
-  ContributionCompleted(ledger::Result::LEDGER_OK, viewing_id);
+    ledger_->UpdateContributionInfoContributedAmount(
+        contribution_id,
+        publisher_key,
+        callback);
+    return;
+  }
+
+  SetTimer(&retry_timers_[contribution_id]);
+}
+
+void Unblinded::OnTokenProcessed(
+    const ledger::Result result,
+    const std::string& contribution_id) {
+  ledger_->GetContributionInfo(
+      contribution_id,
+      std::bind(&Unblinded::CheckIfDone,
+                this,
+                _1));
+}
+
+void Unblinded::CheckIfDone(ledger::ContributionInfoPtr contribution) {
+  if (!contribution) {
+    return;
+  }
+
+  bool completed = true;
+  for (auto& publisher : contribution->publishers) {
+    if (publisher->total_amount == publisher->contributed_amount) {
+      continue;
+    }
+
+    completed = false;
+    break;
+  }
+
+  if (completed) {
+    ContributionCompleted(ledger::Result::LEDGER_OK, std::move(contribution));
+    return;
+  }
+
+  SetTimer(&retry_timers_[contribution->contribution_id]);
```

Comment:
check does retry_timers_ contains a contribution_id element before accessing it.

Reply:
done